xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -729,78 +753,103 @@ absl::Status IrEmitterUnnested::EmitCublasLtMatmulThunk(
                       GetAllocationSliceForHlo(instr->operand(0)));
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice b,
                       GetAllocationSliceForHlo(instr->operand(1)));
   BufferAllocation::Slice c;
-  bool has_matrix_bias = config.beta() != 0;
   if (has_matrix_bias) {
     TF_ASSIGN_OR_RETURN(c, GetAllocationSliceForHlo(instr->operand(2)));
   } else {
     TF_ASSIGN_OR_RETURN(c, GetAllocationSliceForHlo(instr, output_index));
   }
+  int a_scale_index = has_matrix_bias ? 3 : 2;
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice d,
                       GetAllocationSliceForHlo(instr, output_index));
-
-  int a_scale_index = has_matrix_bias ? 3 : 2;
   TF_ASSIGN_OR_RETURN(BufferAllocation::Slice a_scale,
                       GetAllocationSliceForHlo(instr->operand(a_scale_index)));
   TF_ASSIGN_OR_RETURN(
       BufferAllocation::Slice b_scale,
       GetAllocationSliceForHlo(instr->operand(a_scale_index + 1)));
-#if GOOGLE_CUDA
   TF_ASSIGN_OR_RETURN(
       BufferAllocation::Slice c_scale,
       GetAllocationSliceForHlo(instr->operand(a_scale_index + 2)));
   TF_ASSIGN_OR_RETURN(
       BufferAllocation::Slice d_scale,
       GetAllocationSliceForHlo(instr->operand(a_scale_index + 3)));
-#else  // TENSORFLOW_USE_ROCM
-  BufferAllocation::Slice c_scale;
-  BufferAllocation::Slice d_scale;
-#endif
-
-  BufferAllocation::Slice bias;
+  BufferAllocation::Slice aux, bias, d_amax;
   if (has_vector_bias) {
-    TF_ASSIGN_OR_RETURN(
-        bias, GetAllocationSliceForHlo(instr->operand(a_scale_index + 4)));
+    TF_ASSIGN_OR_RETURN(bias, epilogue == GemmBackendConfig::D_RELU_BGRAD
+                                  ? GetAllocationSliceForHlo(instr, {1})
+                                  : GetAllocationSliceForHlo(
+                                        instr->operand(a_scale_index + 4)));
   }
 
-  BufferAllocation::Slice d_amax;
-  if (has_damax) {
-    TF_ASSIGN_OR_RETURN(d_amax, GetAllocationSliceForHlo(instr, {1}));
+  if (!is_relu_epilogue) {
+    bool has_damax = has_tuple_output;
+    if (has_damax) {
+      TF_ASSIGN_OR_RETURN(d_amax, GetAllocationSliceForHlo(instr, {1}));
+    }
+    // aux not used.
+  } else if (epilogue == GemmBackendConfig::D_RELU ||
+             epilogue == GemmBackendConfig::D_RELU_BGRAD) {
+    // cublasLt does not support fp8 output for D_RELU or D_RELU_BGRAD
+    if (has_aux_input) {
+      TF_ASSIGN_OR_RETURN(aux, GetAllocationSliceForHlo(
+                                   instr->operand(instr->operand_count() - 1)));
+    }
+    // d_amax not used.
+  } else if (epilogue == GemmBackendConfig::RELU_AUX ||
+             epilogue == GemmBackendConfig::BIAS_RELU_AUX) {
+    if (has_aux_output) {
+      TF_ASSIGN_OR_RETURN(aux, GetAllocationSliceForHlo(instr, {1}));
+    }
+    bool has_damax = has_tuple_output &&
+                     instr->shape().tuple_shapes().size() >
+                         (epilogue == GemmBackendConfig::RELU_AUX ? 2 : 1);
```

Comment:
Is there valid case with instr the shape of a tuple with one element?

Reply:
I cannot think of one. In this case, it's 3 for RELU_AUX and 2 for BIAS_RELU_AUX.